kubeconfig.write: update the issuer with ingressPort

vyasgun opened the pull request:
**Fixes:** Issue #4124 

**Relates to:** Issue #4177 (_Maybe_)

## Solution/Idea

This is a workaround to explicitly set the `ingressPortNumber` in the `Issuer` field when requesting token. By default, the OpenShift library uses the URL from `/.well-known/oauth-authorization-server` without including the custom port set by the user on CRC. This omission causes errors when writing new contexts to kubeconfig, as the token request fails due to the missing port number. Since users configure this custom port for their CRC instance, this commit ensures the issuer field includes the ingress-https-port before making the token request.
## Proposed changes

List main as well as consequential changes you introduced or had to introduce.

1. Wrote a new equivalent local function to perform the port setting.

## Testing

1. `crc config set ingress-https-port 7443`
2. `crc setup`
3. `crc start`

praveenkumar commented:
Can you put more info in the commit logs why it was not working without this PR and what this PR adds to make it work?

vyasgun committed:
Append ingress-https-port to the issuer field in OAuth token requests

By default, the OpenShift library uses the URL from /.well-known/oauth-authorization-server without including the custom port set by the user on CRC. This omission causes errors when writing new contexts to kubeconfig, as the token request fails due to the missing port number. Since users configure this custom port for their CRC instance, this commit ensures the issuer field includes the ingress-https-port before making the token request.

The default library function SetDefaultOsinConfig fetches OAuth metadata from the /.well-known/oauth-authorization-server endpoint, which returns the issuer URL without the user's custom port number. To prevent token request failures, the new local function requestTokenWithChallengeHandlers appends the custom port to the issuer URL.

vyasgun commented:
Thanks @praveenkumar. Just added more info to the commit log.

praveenkumar commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -198,13 +198,23 @@ func getTokenForUser(username, password, ip string, ca []byte, clusterConfig *ty
+		return "", err
+	}
+
+	o.Issuer = fmt.Sprintf("%s:%d", o.Issuer, port)
```
if we are adding the port explicit to issuer then I don't think we really need https://github.com/crc-org/crc/pull/4183/files#diff-a796f5c0a3b2fdd2f85be66107ff80578a3658029bd675c39ac6c015b515ebecR188-R190 part so let's try to remove that and see if that works.

praveenkumar commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -181,13 +180,10 @@ func getTokenForUser(username, password, ip string, ca []byte, clusterConfig *ty
 				}
-				if strings.HasSuffix(hostname, constants.AppsDomain) {
-					port = strconv.Itoa(int(ingressHTTPSPort))
-				}
```
@vyasgun After removing it looks like it fails again, did you test this scenario?
```
DEBU Using address: api.crc.testing:6443          
DEBU Dialing to 127.0.0.1:6443                    
DEBU Using address: oauth-openshift.apps-crc.testing:443 
DEBU Dialing to 127.0.0.1:443                     
ERRO Cannot update kubeconfig: dial tcp 127.0.0.1:443: connect: connection refused 
DEBU Making call to close driver server           
```

vyasgun replied:
I didn't face the issue. I will take another look and update.

vyasgun replied:
I added the lines back and updated the PR.

anjannath commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -10,7 +10,7 @@ import (
 	"os"
 	"path/filepath"
-	"strconv"
+	`strconv`
```
unrelated change probably by the ide?

praveenkumar replied:
@vyasgun can you fix it and repush it?

gbraad replied:
@vyasgun can you handle this ASAP, otherwise it will not be possible to merge in time for the next release

gbraad replied:
did linting fail? otherwise we can merge regardless and fix this ASAP if not done today.

cfergeau replied:
This PR has " Maintainers are allowed to edit this pull request.", I've fixed this.

vyasgun replied:
thanks @cfergeau! Sorry I was AFK

cfergeau commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -198,13 +198,23 @@ func getTokenForUser(username, password, ip string, ca []byte, clusterConfig *ty
+		return "", err
+	}
+
+	o.Issuer = fmt.Sprintf("%s:%d", o.Issuer, port)
```
Using https://pkg.go.dev/net#JoinHostPort may be slightly more correct. But there is this issue throughout the codebase, all of these can be changed in one go the day we look at proper ipv6 support.

vyasgun replied:
thanks, will change this.

praveenkumar replied:
This was mistake because `o.Issuer` is URL and not the host string and because of that if take it as ip6 and fails
```
ERRO Cannot update kubeconfig: parse "[https://oauth-openshift.apps-crc.testing]:443": first path segment in URL cannot contain colon 
```
